### ISSUE TITLE
Removing packages which replaces another does not remove package and install replacement.

### ISSUE THREAD

**Issue by bendavies** [composer/composer#7573]
My `composer.json`:

```json
{
    "require": {
        "symfony/maker-bundle": "^1.6",
        "phpstan/phpstan-shim": "^0.10.3"
    }
}
```

Output of `composer diagnose`:

```
Checking composer.json: WARNING
No license specified, it is recommended to do so. For closed-source software you may use "proprietary" as license.
Checking platform settings: OK
Checking git settings: OK
Checking http connectivity to packagist: OK
Checking https connectivity to packagist: OK
Checking github.com oauth access: OK
Checking disk free space: OK
Checking pubkeys:
Tags Public Key Fingerprint: 57815BA2 7E54DC31 7ECC7CC5 573090D0  87719BA6 8F3BB723 4E5D42D0 84A14642
Dev Public Key Fingerprint: 4AC45767 E5EC2265 2F0C1167 CBBB8A2B  0C708369 153E328C AD90147D AFE50952
OK
Checking composer version: OK
Composer version: 1.7.2
PHP version: 7.2.7
PHP binary path: /usr/local/Cellar/php/7.2.7/bin/php
```

When I run this command:

```
composer remove phpstan/phpstan-shim
```

I get the following output:

```
Loading composer repositories with package information
Updating dependencies (including require-dev)
Nothing to install or update
Writing lock file
Generating autoload files
No changelogs summary
```

And I expected this to happen:

`phpstan/phpstan-shim` to be removed and `nikic/php-parser` to be installed.

**Comment by alcohol** [comment]
That is probably a bit complex to fix easily. The solver basically detects that there is still a requirement, and there is currently a package fulfilling that requirement. Since you did not explicitly require `nikic/php-parser`, there is no reason to replace `phpstan/phpstan-shim` with `nikic/php-parser`.

**Comment by alcohol** [comment]
It becomes more problematic though once you explicitly have a root-requirement for `nikic/php-parser`:

```
rob@galaga /tmp/test $ composer require nikic/php-parser
Using version ^4.0 for nikic/php-parser
./composer.json has been updated
Loading composer repositories with package information
Updating dependencies (including require-dev)
Nothing to install or update
Writing lock file
Generating autoload files

rob@galaga /tmp/test $ composer update
Loading composer repositories with package information
Updating dependencies (including require-dev)
Nothing to install or update
Generating autoload files
```

:-/

**Comment by naderman** [comment]
I'm really starting to think we just need to get rid of anything from the vendor folder altogether during dependency resolution. It may have been a nice idea to avoid unnecessary work, but I think reproducible results are a lot more important, but probably something to fix for Composer v2 /cc @Seldaek

**Comment by alcohol** [comment]
Or maybe just not take it into consideration at least? Then once solving is complete, attempt to compare state against solving solution and attempt to resolve the diff?

**Comment by naderman** [comment]
@alcohol that's what I meant by getting rid of it during dep resolution, actual execution of the resulting transaction after dep resolution can obviously keep whatever is present

**Comment by Seldaek** [comment]
Fixed already in 2.0
